src/hotspot/share/gc/parallel/psCardTable.cpp

```diff
@@ -123,70 +124,166 @@ static void prefetch_write(void *p) {
+  assert(start < end, "precondition");
+  assert(is_card_aligned(start), "precondition");
+
+  // Helper struct to keep the following code compact.
+  struct Obj {
+    HeapWord* addr;
+    oop obj;
+    bool is_obj_array;
+    HeapWord* end_addr;
+    Obj(HeapWord* o_addr) : addr(o_addr),
+                            obj(cast_to_oop(o_addr)),
+                            is_obj_array(obj->is_objArray()),
+                            end_addr(addr + obj->size()) {}
+    void next() {
+      addr = end_addr;
+      obj = cast_to_oop(addr);
+      is_obj_array = obj->is_objArray();
+      end_addr = addr + obj->size();
     }
-    assert(i_card - 1 >= start_card, "inv");
-    assert(*(i_card - 1) != PSCardTable::clean_card_val(), "prev card must be dirty");
-    // Find the final obj on the prev dirty card.
-    HeapWord* obj_addr = start_array->object_start(addr_for(i_card)-1);
-    HeapWord* obj_end_addr = obj_addr + cast_to_oop(obj_addr)->size();
-    CardValue* final_card_by_obj = byte_for(obj_end_addr - 1);
-    assert(final_card_by_obj < end_card, "inv");
-    if (final_card_by_obj <= i_card) {
-      return i_card;
+  };
+
+  StripeShadowTable sct(this, MemRegion(start, end));
+
+  // end might not be card-aligned
+  const CardValue* end_card = sct.card_for(end - 1) + 1;
+
+  for (HeapWord* i_addr = start; i_addr < end; /* empty */) {
+    const CardValue* dirty_l = sct.find_first_dirty_card(sct.card_for(i_addr), end_card);
+    const CardValue* dirty_r = sct.find_first_clean_card(dirty_l, end_card);
+
+    assert(dirty_l <= dirty_r, "inv");
+
+    if (dirty_l == dirty_r) {
+      assert(dirty_r == end_card, "inv");
+      break;
     }
-    // This final obj extends beyond i_card, check if this new card is dirty.
-    if (*final_card_by_obj == PSCardTable::clean_card_val()) {
-      return final_card_by_obj;
+
+    // Located a non-empty dirty chunk [dirty_l, dirty_r)
+    HeapWord* addr_l = sct.addr_for(dirty_l);
+    HeapWord* addr_r = MIN2(sct.addr_for(dirty_r), end);
+
+    // Scan objects overlapping [addr_l, addr_r) limited to [start, end)
+    Obj obj(object_start(addr_l));
+
+    while (true) {
+      assert(obj.addr < addr_r, "inv");
+
+      if (obj.is_obj_array) {
+        // precise-marked
+        scan_obj_with_limit(pm, obj.obj, addr_l, addr_r);
+      } else {
+        if (obj.addr < addr_l) {
+          if (sct.is_any_dirty(sct.card_for(MAX2(obj.addr, start)), dirty_l)) {
+            // already scanned
+          } else {
+            // treat it as semi-precise-marked, [addr_l, obj-end)
+            scan_obj_with_limit(pm, obj.obj, addr_l, MIN2(obj.end_addr, end));
+          }
+        } else {
+          // obj-start is dirty
+          if (obj.end_addr <= end) {
+            // scan whole obj if it does not extend beyond the stripe end
+            scan_obj(pm, obj.obj);
+          } else {
+            // otherwise scan limit the scan
+            scan_obj_with_limit(pm, obj.obj, addr_l, end);
+          }
+        }
```

Comment:
I wonder if the else branch can be replaced by:

```c
  if (obj.addr < i_addr && obj.addr > start) {
    // already-scanned
  } else {
    scan_obj_with_limit(pm, obj.obj, addr_l, end);
  }
```

Reply:
You mean to replace L220-L227, right? There we know `obj.addr >= addr_l`. `obj.addr < i_addr` cannot be true then IMO. It would overlap with an objArray if `i_addr` was set there. If `i_addr` was set at the end of a non-objArray then objects starting before `i_addr` cannot start in or after `addr_l`.

Reply:
No, I mean L212 to L228.  "Comment on lines +212 to +228"

Reply:
Yes indeed. Looks correct to me. Nice simplification!

Reply:
> No, I mean L212 to L228. "Comment on lines +212 to +228"

I looked at this at first but thought it is obviously wrong if `obj` is precisely marked after a previous young collection. But if it is precisely marked it is of course correct scan from the beginning of the dirty chunk (addr_l) to the obj/stripe end.

Reply:
I  get `assert(should_scavenge(p, true)) failed: revisiting object?` with test/jdk/java/lang/Thread/virtual/stress/Skynet.java when the change is applied.

It reproduces well but not 100%. Need to look into it.

Reply:
`obj` can be scanned twice if it reaches into the stripe because `obj.addr > start` will still false upon the 2nd time. Fix is to check `i_addr > start`.